### PR TITLE
8180892: Correct handling of annotations on parameters

### PR DESCRIPTION
**Pull request by liach**
This patch aims to correct handling of annotations on parameters with the help of `MethodParameters` attribute, which will be always available once #9862 is integrated.

It utilizes and expands upon the existing parameter matching logic present in `Executable::getAllGenericParameterTypes`, and delegate parameter parameterized types, parameter annotation, and parameter type annotation accesses through the matched results, if matching is available. If matching failed, it falls back to existing heuristics that works without `MethodParameters` attributes for annotations, in `Executable::handleParameterNumberMismatch` and `TypeAnnotationParser::buildAnnotatedTypes` (renamed `buildAnnotatedTypesWithHeuristics` in this patch)

`ParameterMappingTest` covers these scenarios with class files that have `MethodParameters` or `Signature` attributes stripped or preserved to ensure the new Reflection API implementation works for both class files generated before #9862 and after its integration.

Also special thanks to Joe Darcy for reviewing 8304918 (#13183); this brings much convenience to this patch.

---------
### Progress
- [ ] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [ ] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8180892](https://bugs.openjdk.org/browse/JDK-8180892): Correct handling of annotations on parameters (**Bug** - P4)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/13664/head:pull/13664` \
`$ git checkout pull/13664`

Update a local copy of the PR: \
`$ git checkout pull/13664` \
`$ git pull https://git.openjdk.org/jdk.git pull/13664/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 13664`

View PR using the GUI difftool: \
`$ git pr show -t 13664`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/13664.diff">https://git.openjdk.org/jdk/pull/13664.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/13664#issuecomment-1522780288)

### CODE REVIEW

**Comment by jddarcy** [comment]
@liach , thank you for working on this fix. I've given the changes an initial skim, but will need to do a more thorough review before approving or offering suggestions for changes. My top priority for the next few weeks will be JDK 21 CSRs so it may take a while for me to circle back to this PR.

**Comment by liach** [comment]
Now that JDK 21 is RDP 1, should we take a look at this now?
This patch makes use of MethodParameters information to map (explicit) parameter annotations to formal parameters. In case that information is not available, the code falls back to the existing heuristic that guesses the offset for inner classes.

Two more potential issues with core reflection and annotations/generics include:
1. `getGenericParameterTypes()[i]` usages should be replaced by `getParameters()[i].getParameterizedType()`, which returns the correct generic type if MethodParameters attribute is present
2. There is some issues with `Constructor.getAnnotatedReceiverType()` due to some effectively-static local and anonymous classes (like ones in static methods and initializers); this is tracked at [JDK-8162500](https://bugs.openjdk.org/browse/JDK-8162500). Those classes don't carry enough information about whether they should have an mandated outer instance parameter, such as through a static flag in InnerClasses attribute; might need Javac tweaks to fix that.

File: src/java.base/share/classes/java/lang/reflect/Executable.java [hunk 1]
```diff
@@ -441,20 +400,92 @@ private ParameterData parameterData() {
...
+
+    /**
+     * Converts information from explicit parameters to information about all parameters,
+     * using information from MethodParameters attribute. Guard calls checks to
+     * {@code matchedParameterCount() == explicit.length}.
+     *
+     * @param <T> the type of information
+     * @param explicit information from explicit parameters
+     * @param implicitMapper generates information for implicit parameters
+     * @param freshCopy if true, will copy explicit information on return to avoid leaks
```
Comment (bjkail) [inline src/java.base/share/classes/java/lang/reflect/Executable.java:453]: Is the `freshCopy` parameter needed?

**Comment by liach** [comment]
Keep alive.

**Comment by liach** [comment]
@jddarcy Joe, can you take a look at this patch where annotation handling on parameters now uses available `MethodParameters` attribute to match from explicit formal parameter indices to implicit formal parameter indices?

**Comment by jddarcy** [comment]
> @jddarcy Joe, can you take a look at this patch where annotation handling on parameters now uses available `MethodParameters` attribute to match from explicit formal parameter indices to implicit formal parameter indices?

@liach sorry for the delays.

I'll be attending to some other matters in the immediate future; please ping me again in a few weeks if I don't reply first.

**Comment by liach** [comment]
Hello Joe @jddarcy, is there anything that I need to do to move this patch forward? Should I file a CSR for the behavior change to `getAnnotatedParameterTypes()` and `getParameters()[i].getAnnotatedType()` due to this bugfix, as it now annotates correctly?

**Comment by jddarcy** [comment]
> Hello Joe @jddarcy, is there anything that I need to do to move this patch forward? Should I file a CSR for the behavior change to `getAnnotatedParameterTypes()` and `getParameters()[i].getAnnotatedType()` due to this bugfix, as it now annotates correctly?

Hi @liach , sorry for the continued delays. I'm currently working through reviewing a backlog of CSRs for JDK 22. This is (still) on my list to get to.

File: src/java.base/share/classes/java/lang/reflect/Executable.java (hunk 1 above)
Comment (liach) [reply src/java.base/share/classes/java/lang/reflect/Executable.java]: Removed unused parameter.

**Comment by liach** [comment]
This patch has been updated to the latest Class-File API. May I request a review again?

**Comment by liach** [comment]
Keep-alive.

**Pull request closed without merging**